### DISCUSSION TITLE
Turn off in-app update mechanism

### DISCUSSION THREAD

**Discussion by np25071984** [Homebrew/discussions#4849] [Casks] [answered]
### Output of `brew config`

```shell
$ brew config
HOMEBREW_VERSION: 4.1.15
ORIGIN: https://github.com/Homebrew/brew
HEAD: e57a87cfe9c15f192cd4fc6695a2b3b7acf0ddf7
Last commit: 9 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 6ad0c890991715b8bab11ddc5f3c48b483503a6a
Core tap last commit: 3 days ago
Core tap branch: master
Core tap JSON: 16 Oct 00:40 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.5.1-arm64
CLT: 14.3.1.0.1.1683849156
Xcode: 15.0
Rosetta 2: false
```

### Output of `brew doctor`

```shell
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libDellMonitorSdkLib.dylib

Warning: You have an unnecessary local Cask tap.
This can cause problems installing up-to-date casks.
Please remove it by running:
  brew untap homebrew/cask

Warning: You have an unnecessary local Core tap!
This can cause problems installing up-to-date formulae.
Please remove it by running:
 brew untap homebrew/core
```

### Description of issue

I like to keep my software updated but I don't like to spend much time on it. On Linux I just run `apt upgrade` once a week and get everything fresh and secured. The same I would like to do on MacOS - update all application once a week with `brew upgrade --greedy` The problem is that when there is an update the applications bother me with "Update" button. In Chrome it is located by the search bar, it Telegram is is floats above my chat list. Don't like this much!

Since I update my apps via command line I don't need those notifications at all. So, is there a way how to turn off in-app update feature?

### Relevant casks

google-chrome, telegram

**Comment by SMillerDev** [comment]
No, although you can run `brew upgrade --greedy` to also upgrade those

**Reply by np25071984** [reply]
Does anybody know where those "Internal update" feature were added to software? In order to get rid of those "Update" buttons whom I should talk to?

**Reply by SMillerDev** [reply]
To whoever made the software

**Reply by np25071984** [reply]
And why don't they do this on Linux OS?

**Reply by SMillerDev** [reply]
They definitely do, casks don't exist on Linux though so maybe that's where your confusion comes from.

**Reply by np25071984** [reply]
No, they definitely don't. I have been using Linux for decades and have never seen "Update" buttons in such annoying manner on Linux (hasn't used brew though) This is my biggest pain since I have to use MacOS - in order to update my software I have to open up all my installed casks, find "check for update" button, hit it and wait. For EACH cask entities! Isn't it insane? Furthermore, this "Update" button can pop up on Monday, for example, and hang around whole week long until I am ready to do update my system (usually it is Saturday).
I have tons video calls and I may say all my teammates who uses MacOS has the same issue - their software is usually outdated and there are those "Update me" buttons around. It is hilarious )
On Linux all I get - text notification when I logged in the shell that there are security updates available. That is it! Package manager may notify me in the system tray but I am able to turn it off at any time. It is handy!

**Reply by SMillerDev** [reply]
You seem to be confusing brew updates with app updates here. Your package manager on Ubuntu will notify you on terminal login, but apps like Chrome show "update" buttons all the same between each OS.
The main difference is that brew doesn't update software that auto-updates by default.

Either way, the only ones who can change the update mechanism of an app are the developers.

**Reply by np25071984** [reply]
Ok, let's check it out? Here is my laptop I haven't been using it for awhile. Ubuntu OS with `apt` package manager on it.

```
$ apt list -a google-chrome-stable
Listing... Done
google-chrome-stable/stable 119.0.6045.159-1 amd64 [upgradable from: 119.0.6045.123-1]
google-chrome-stable/now 119.0.6045.123-1 amd64 [installed,upgradable to: 119.0.6045.159-1]
```

We can see my Google Chrome version is quite obsolete (it is `119.0.6045.123-1` vs `119.0.6045.159-1` the last available). The package manager offers to update it. I don't know how many version there are between my current and the last one though.

And this is how it looks like if I open the app:
![image](https://github.com/Homebrew/discussions/assets/19670887/3665fb24-8fd9-4d72-999c-a3ecb24e4426)
No any "Upgrade" buttons or notificatinos at all!

**Reply by np25071984** [reply]
Furthermore, I have also tried snap package manager. The same situation - applications which were installed with it don't have such "internal upgrade mechanism" - all I need just run `snap refresh` and that is it!

**Reply by np25071984** [reply]
```
$ snap info firefox
name:      firefox
summary:   Mozilla Firefox web browser
publisher: Mozilla✓
store-url: https://snapcraft.io/firefox
contact:   https://support.mozilla.org/kb/file-bug-report-or-feature-request-mozilla
license:   unset
description: |
  Firefox is a powerful, extensible web browser with support for modern web application
  technologies.
commands:
  - firefox
  - firefox.geckodriver
snap-id:      3wdHCAVyZEmYsCMFDE9qt92UV8rC8Wdk
tracking:     latest/stable
refresh-date: 10 days ago, at 18:38 EST
channels:
  latest/stable:    120.0-2      2023-11-21 (3416) 257MB -
  latest/candidate: 120.0-2      2023-11-16 (3416) 257MB -
  latest/beta:      121.0b2-1    2023-11-22 (3451) 258MB -
  latest/edge:      122.0a1      2023-11-22 (3447) 291MB -
  esr/stable:       115.5.0esr-1 2023-11-21 (3392) 253MB -
  esr/candidate:    115.5.0esr-1 2023-11-14 (3392) 253MB -
  esr/beta:         ↑                                    
  esr/edge:         ↑                                    
installed:          119.0.1-1               (3358) 251MB -
```

![image](https://github.com/Homebrew/discussions/assets/19670887/44b12091-7523-4c14-b8f3-bb0e4ccaa332)

**Reply by alebcay** [reply, marked as answer]
If your Chrome is coming from apt repository `http://dl.google.com/linux/chrome/deb/`, then presumably Google is building Chrome without auto-update support. In other words, Google has designated your package manager as the update mechanism and blessed this approach, turning off auto-updates within the application itself.

For the Firefox snap, it is also provided by Mozilla (Mozilla is the publisher). Mozilla is presumably building Firefox without enabling the updater, since they plan to ship new updates via the snap mechanism.

---

Over here in Homebrew Cask, we only have access to the binaries distributed by Google and Mozilla for their respective browsers for macOS. These binaries have auto-update support enabled, because they expect most users on macOS to install the browser by downloading and installing it without a package manager, and thus it needs to have some way to update itself.

We have no easy way of disabling this mechanism, short of perhaps building the browsers from source (which we can't even do with Chrome anyways) with the appropriate patches to disable the update options.

Basically - what users can obtain through Homebrew Cask is no different from what can be obtained directly from the website for macOS users. And Google and Mozilla have both decided that macOS users must be subjected to automatic updates, short of whatever configuration may be exposed in the applications to end users to disable it.

---

If you can convince Google and/or Mozilla to make available a binary distribution of each version of their browsers, with the update mechanism disabled (similar to what Mozilla is doing with snaps), we can proceed to distribute those via Homebrew Cask. But so far, most vendors for software on macOS prefer to have the auto-update function built-in, due to the lack of a built-in package manager that they would rely on.

**Reply by np25071984** [reply]
> because they expect most users on macOS to install the browser by downloading and installing it without a package manager, and thus it needs to have some way to update itself.

Makes perfect sense!

@alebcay thank you for your answer, not it is way more clear for me.
